File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            res[i] = 1;
        }
        
        int prefix = 1;
        int posfix = 1;

        for (int i = 0; i < n; i++) {
            res[i] = prefix;
            prefix *= nums[i];
        }
        for (int j = n - 1; j >= 0; j--) {
            res[j] *= posfix;
            posfix *= nums[j];
        }
        return res;
    }
}  

File: Data Structures & Algorithms/combination-target-sum/submission-5.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        # optimal way
        res = []
        nums.sort()

        def dfs(i, cur, total):
            if total == target:
                res.append(cur[:])
                return res
            
            for j in range(i, len(nums)):
                if total + nums[j] > target:
                    return 
                cur.append(nums[j])
                dfs(j, cur, total + nums[j])
                cur.pop()
        
        dfs(0, [], 0)
        return res

        # Time complexity: O(t / 2m) where t is target and m is minimum value in nums
        # Space complexity: O(t / m)